### PR TITLE
Prefetch server implementation 

### PR DESCRIPTION
**Pull request by tgraham89**
First implementation of prefetching functionality

### CODE REVIEW

File: rangeserver/src/error.rs [hunk 1]
```diff
@@ -53,6 +54,7 @@ impl Error {
             Self::UnknownTransaction => Status::UnknownTransaction,
             Self::TransactionAborted(_) => Status::TransactionAborted,
             Self::InternalError(_) => Status::InternalError,
+            Self::PrefetchError => Status::PrefetchError,
```
Comment (tamer-eldeeb) [inline rangeserver/src/error.rs:57]: Could you implement the other direction (in the function from_flatbuf_status)

File: rangeserver/src/lib.rs [hunk 2]
```diff
@@ -2,6 +2,7 @@ pub mod epoch_provider;
 pub mod error;
 pub mod for_testing;
 mod key_version;
+pub mod prefetching_buffer;
```
Comment (tamer-eldeeb) [inline rangeserver/src/lib.rs:5]: minor: does this need to be pub?

File: rangeserver/src/range_manager.rs [hunk 3]
```diff
@@ -564,10 +586,64 @@ where
...
+            .storage
+            .get(self.range_id, key.clone())
+            .await
+            .map_err(Error::from_storage_error)?;
+        Ok(val)
+    }
+
+    pub async fn process_prefetch(
+        &self,
+        buffer: &Arc<PrefetchingBuffer>,
```
Comment (tamer-eldeeb) [inline rangeserver/src/range_manager.rs:611]: the range manager already has a reference to the prefetching buffer, so this is probably not necessary?

File: rangeserver/src/range_manager.rs [hunk 4]
```diff
@@ -564,10 +586,64 @@ where
...
+    pub async fn prefetch_get(&self, key: Bytes) -> Result<Option<Bytes>, Error> {
+        let val = self
+            .storage
+            .get(self.range_id, key.clone())
+            .await
+            .map_err(Error::from_storage_error)?;
+        Ok(val)
+    }
+
+    pub async fn process_prefetch(
```
Comment (tamer-eldeeb) [inline rangeserver/src/range_manager.rs:609]: I'd just call this "prefetch", in line with the other function names

**Review by tamer-eldeeb — CHANGES_REQUESTED** [review]

File: rangeserver/src/range_manager.rs [hunk 5]
```diff
@@ -484,6 +491,11 @@ where
                         .map_err(Error::from_wal_error)?;
                 }
                 lock_table.release();
+                // Call transaction_completed for prebuffer
```
Comment (tamer-eldeeb) [inline rangeserver/src/range_manager.rs:494]: nit: comments that simply state what the line below does are not useful :)

File: rangeserver/src/range_manager.rs [hunk 6]
```diff
@@ -544,18 +556,28 @@ where
...
 
                             self.storage
-                                .upsert(self.range_id, key, val, version)
+                                .upsert(self.range_id, key.clone(), val.clone(), version)
                                 .await
                                 .map_err(Error::from_storage_error)?;
+
+                            // Update the prefetch buffer if this key has been requested by a prefetch call
+                            self.prefetching_buffer
+                                .process_buffer_update(key, val)
```
Comment (tamer-eldeeb) [inline rangeserver/src/range_manager.rs:565]: I think this also should just be simply called "upsert"

File: rangeserver/src/range_manager.rs [hunk 7]
```diff
@@ -544,18 +556,28 @@ where
...
                             let key = Bytes::copy_from_slice(del.k().unwrap().bytes());
                             self.storage
-                                .delete(self.range_id, key, version)
+                                .delete(self.range_id, key.clone(), version)
                                 .await
                                 .map_err(Error::from_storage_error)?;
+
+                            // Delete the key from the prefetch buffer if this key has been requested by a prefetch call
+                            self.prefetching_buffer
+                                .process_transaction_delete(key)
```
Comment (tamer-eldeeb) [inline rangeserver/src/range_manager.rs:579]: similarly, just call this function delete

File: rangeserver/src/server.rs [hunk 8]
```diff
@@ -1,11 +1,12 @@
 use bytes::Bytes;
 use common::network::fast_network::FastNetwork;
-use std::collections::HashMap;
+use std::collections::{BTreeMap, HashMap};
```
Comment (tamer-eldeeb) [inline rangeserver/src/server.rs:3]: is this still used here?

File: rangeserver/src/server.rs [hunk 9]
```diff
@@ -26,23 +27,73 @@ use flatbuf::rangeserver_flatbuffers::range_server::*;
 use proto::rangeserver::range_server_server::{RangeServer, RangeServerServer};
 use proto::rangeserver::{PrefetchRequest, PrefetchResponse};
 
-#[derive(Debug, Default)]
-struct ProtoServer {}
+use crate::prefetching_buffer::PrefetchingBuffer;
+
+pub mod rangeserver {
```
Comment (tamer-eldeeb) [inline rangeserver/src/server.rs:32]: why is this needed?

File: rangeserver/src/server.rs [hunk 10]
```diff
@@ -26,23 +27,73 @@ use flatbuf::rangeserver_flatbuffers::range_server::*;
...
+    include!("../../proto/target/rangeserver/rangeserver.rs");
+}
+
+#[derive(Clone)]
+struct ProtoServer<S, E>
+where
+    S: Storage,
+    E: EpochProvider,
+{
+    buffer: Arc<PrefetchingBuffer>,
```
Comment (tamer-eldeeb) [inline rangeserver/src/server.rs:42]: don't think we need that reference here.

File: rangeserver/src/server.rs [hunk 11]
```diff
@@ -651,7 +705,13 @@ where
             .parse()
             .unwrap();
 
-        let prefetch = ProtoServer::default();
+        // Create buffer to hold prefetch requests
+        // let prefetching_buffer = PrefetchingBuffer::new();
```
Comment (tamer-eldeeb) [inline rangeserver/src/server.rs:709]: please remove commented out code

File: rangeserver/src/server.rs [hunk 12]
```diff
@@ -651,7 +705,13 @@ where
             .parse()
             .unwrap();
 
-        let prefetch = ProtoServer::default();
+        // Create buffer to hold prefetch requests
```
Comment (tamer-eldeeb) [inline rangeserver/src/server.rs:708]: same as above :)

File: rangeserver/src/server.rs [hunk 13]
```diff
@@ -26,23 +27,73 @@ use flatbuf::rangeserver_flatbuffers::range_server::*;
...
         };
 
-        Ok(Response::new(reply)) // Send back our formatted greeting
+        let range_manager = self
+            .parent_server
+            .maybe_load_and_get_range(&full_range_id)
+            .await
+            .map_err(|e| TStatus::internal(format!("Failed to load range: {:?}", e)))?;
+
+        // Call process_prefetch_request in range_manager.rs
```
Comment (tamer-eldeeb) [inline rangeserver/src/server.rs:87]: another redundant comment :)

File: rangeserver/src/range_manager.rs [hunk 14]
```diff
@@ -564,10 +586,60 @@ where
...
+            .await
+            .map_err(Error::from_storage_error)?;
+        Ok(val)
+    }
+
+    pub async fn prefetch(&self, transaction_id: Uuid, key: Bytes) -> Result<(), Error> {
+        // Request prefetch from the prefetching buffer
+        if let Some(keystate) = self
+            .prefetching_buffer
+            .process_prefetch_request(transaction_id, key.clone())
```
Comment (tamer-eldeeb) [inline rangeserver/src/range_manager.rs:613]: I haven't yet looked into the prefetching buffer module itself, but why does this return an option?

**Review by tamer-eldeeb — CHANGES_REQUESTED** [review]

File: rangeserver/src/range_manager.rs (hunk 14 above)
Comment (tgraham89) [reply rangeserver/src/range_manager.rs:609]: Originally it was written to be able to return None if there was a failure. I have updated to just return the keystate status directly given the function always returns the key as status requested at a minimum.

File: rangeserver/src/server.rs (hunk 9 above)
Comment (tamer-eldeeb) [reply rangeserver/src/server.rs:32]: I was referring to the rangeserver include, I am able to build despite removing it.

File: rangeserver/src/prefetching_buffer.rs [hunk 15]
```diff
@@ -0,0 +1,946 @@
...
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum KeyState {
+    Fetched,        // Key has been fetched and is in the BTree
+    Loading(u64),   // Key has been requested and is being fetched
+    Requested(u64), // Key has been requested but fetching has not been started
+}
+
+#[derive(Default, Debug)]
+pub struct PrefetchingBuffer {
+    pub prefetch_store: Arc<Mutex<BTreeMap<Bytes, Bytes>>>, // stores key / value
```
Comment (tamer-eldeeb) [inline rangeserver/src/prefetching_buffer.rs:18]: I suspect there are way too many mutexes here. These are not free performance wise, but also complicate the code since you have to be careful to lock them in the right order to avoid deadlocks.

Is it possible to create a struct(s) to store the state you need on, and then wrap it in one/few mutexes instead?

File: rangeserver/src/prefetching_buffer.rs [hunk 16]
```diff
@@ -0,0 +1,946 @@
...
+    pub key_state: Arc<Mutex<HashMap<Bytes, KeyState>>>,    // stores key -> current fetch state
+    pub transaction_keys: Arc<Mutex<HashMap<Uuid, BTreeSet<Bytes>>>>, // stores transaction_id -> set of requested keys
+    pub key_transactions: Arc<Mutex<HashMap<Bytes, BTreeSet<Uuid>>>>, // stores key -> set of transaction_ids
+    pub key_state_watcher: Arc<Mutex<HashMap<Bytes, watch::Receiver<KeyState>>>>, // key -> receiver for state changes
+    pub key_state_sender: Arc<Mutex<HashMap<Bytes, watch::Sender<KeyState>>>>, // key -> sender for state changes
+    pub fetch_sequence_number: Mutex<u64>,
```
Comment (tamer-eldeeb) [inline rangeserver/src/prefetching_buffer.rs:24]: A mutex is probably fine, but consider using an atomic here:
https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU64.html#method.fetch_add

File: rangeserver/src/prefetching_buffer.rs [hunk 17]
```diff
@@ -0,0 +1,946 @@
...
+            let _ = prefetch_store.remove(&key); // might be None
+            let _ = key_state.remove(&key); // might be None
+        }
+    }
+
+    /// If a transaction deletes a key, it must also be deleted from the buffer
+    /// This function deletes a key from prefetch_store and key_state
+    pub async fn delete(&self, key: Bytes) {
+        let mut prefetch_store = self.prefetch_store.lock().await;
+        let mut key_state = self.key_state.lock().await;
```
Comment (tamer-eldeeb) [inline rangeserver/src/prefetching_buffer.rs:276]: we should not remove the key_state for a deleted key! we want to maintain in memory the fact that this key does not exist, so that a transaction does not need to go to storage to learn that fact.

**Review by tamer-eldeeb — CHANGES_REQUESTED** [review]

File: rangeserver/src/server.rs (hunk 9 above)
Comment (tgraham89) [reply rangeserver/src/server.rs:32]: Oh I see, you're right.

File: rangeserver/src/prefetching_buffer.rs (hunk 17 above)
Comment (tgraham89) [reply rangeserver/src/prefetching_buffer.rs:283]: I added a KeyState::Deleted to maintain this state. We should discuss the key_state logic if a value for a key is reinserted after deletion.

File: rangeserver/src/prefetching_buffer.rs [hunk 18]
```diff
@@ -0,0 +1,968 @@
...
+    /// This function deletes a key from prefetch_store
+    pub async fn delete(&self, key: Bytes) {
+        let mut cur_state = self.state.lock().await;
+        let _ = cur_state.prefetch_store.remove(&key); // might be None
+    }
+
+    /// If no transactions are  requesting a key, it can be flushed from the buffer
+    /// This function deletes a key from prefetch_store and the key_state
+    /// It needs be called with the lock held
+    async fn transaction_delete(&self, key: Bytes, state: &mut State) {
```
Comment (tamer-eldeeb) [inline rangeserver/src/prefetching_buffer.rs:275]: that name is a bit confusing. Maybe evict_key?

File: rangeserver/src/prefetching_buffer.rs [hunk 19]
```diff
@@ -0,0 +1,968 @@
...
+                cur_state.key_state.insert(key.clone(), KeyState::Fetched);
+                // Notify all watchers of the state change
+                if let Some(sender) = cur_state.key_state_sender.get(&key) {
+                    let _ = sender.send(KeyState::Fetched);
+                }
+            }
+        } else {
+            // If key is not being requested by any transactions, remove from prefetch_store instead of update
+            let _ = cur_state.prefetch_store.remove(&key); // might be None
+            let _ = cur_state.key_state.remove(&key); // might be None
```
Comment (tamer-eldeeb) [inline rangeserver/src/prefetching_buffer.rs:261]: you already have a function that does these two lines. Also, I don't think doing it during upsert is very useful.
I would change the structure of this function slightly: check if the key is in the key_state map (in whatever state), then do the transition to Fetched and notification, and don't worry about evicting it here since it's done after the transaction completes.

**Review by tamer-eldeeb — CHANGES_REQUESTED** [review]

File: rangeserver/src/prefetching_buffer.rs [hunk 20]
```diff
@@ -0,0 +1,968 @@
...
+                cur_state.key_state.insert(key.clone(), KeyState::Fetched);
+                // Notify all watchers of the state change
+                if let Some(sender) = cur_state.key_state_sender.get(&key) {
+                    let _ = sender.send(KeyState::Fetched);
+                }
+            }
+        }
+    }
+
+    pub async fn fetch_failed(&self, key: Bytes, fetch_sequence_number: u64) {
```
[the remark this answers is not in the document]
Comment (tgraham89) [reply rangeserver/src/prefetching_buffer.rs:210]: I think I see what you're saying. Updated.

**Review by tamer-eldeeb — APPROVED** [review]
